algopy_testing/src/algopy/op.py

```diff
@@ -168,3 +178,398 @@ def vrf_verify(
     raise NotImplementedError(
         "'op.vrf_verify' is not implemented. Mock using preferred testing tools."
     )
+
+
+def addw(a: UInt64 | int, b: UInt64 | int, /) -> tuple[UInt64, UInt64]:
+    a = as_int64(a)
+    b = as_int64(b)
+    result = a + b
+    return _int_to_uint128(result)
+
+
+def base64_decode(e: Base64, a: Bytes | bytes, /) -> Bytes:
+    a_str = _bytes_to_string(a, "illegal base64 data")
+    a_str = a_str + "="  # append padding to ensure there is at least one
+
+    result = (
+        base64.urlsafe_b64decode(a_str) if e == Base64.URLEncoding else base64.b64decode(a_str)
+    )
+    return Bytes(result)
+
+
+def bitlen(a: Bytes | UInt64 | bytes | int, /) -> UInt64:
+    int_value = int.from_bytes(as_bytes(a)) if (isinstance(a, Bytes | bytes)) else as_int64(a)
+    return UInt64(int_value.bit_length())
+
+
+def bsqrt(a: BigUInt | int, /) -> BigUInt:
+    a = as_int512(a)
+    return BigUInt(math.isqrt(a))
+
+
+def btoi(a: Bytes | bytes, /) -> UInt64:
+    a_bytes = as_bytes(a)
+    if len(a_bytes) > 8:
+        raise ValueError(f"btoi arg too long, got [{len(a_bytes)}]bytes")
+    return UInt64(int.from_bytes(a_bytes))
+
+
+def bzero(a: UInt64 | int, /) -> Bytes:
+    a = as_int64(a)
+    if a > MAX_BYTES_SIZE:
+        raise ValueError("bzero attempted to create a too large string")
+    return Bytes(b"\x00" * a)
+
+
+def divmodw(
+    a: UInt64 | int, b: UInt64 | int, c: UInt64 | int, d: UInt64 | int, /
+) -> tuple[UInt64, UInt64, UInt64, UInt64]:
+    i = _uint128_to_int(a, b)
+    j = _uint128_to_int(c, d)
+    d = i // j
+    m = i % j
+    return _int_to_uint128(d) + _int_to_uint128(m)
+
+
+def divw(a: UInt64 | int, b: UInt64 | int, c: UInt64 | int, /) -> UInt64:
+    i = _uint128_to_int(a, b)
+    c = as_int64(c)
+    return UInt64(i // c)
+
+
+def err() -> None:
+    raise ArithmeticError("err opcode executed")
+
+
+def exp(a: UInt64 | int, b: UInt64 | int, /) -> UInt64:
+    a = as_int64(a)
+    b = as_int64(b)
+    if a == b and a == 0:
+        raise ArithmeticError("0^0 is undefined")
+    return UInt64(a**b)
+
+
+def expw(a: UInt64 | int, b: UInt64 | int, /) -> tuple[UInt64, UInt64]:
+    a = as_int64(a)
+    b = as_int64(b)
+    if a == b and a == 0:
+        raise ArithmeticError("0^0 is undefined")
+    result = a**b
+    return _int_to_uint128(result)
+
+
+def extract(a: Bytes | bytes, b: UInt64 | int, c: UInt64 | int, /) -> Bytes:
```

Comment:
There is a special edge case we need to handle here: 
When b and c are are both `int` (and only `int`) and b <= 255, and c == `0` then the c is treated as to the end of the bytes array.

See https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/v10/#extract

Reply:
It seems that `op.extract` is compiled down to `extract3` in teal which does not seem to behave the same as `extract` op code when `c == 0`. In the tests, I got empty string for the test case where both `a` and `b` are `0`. 

https://github.com/algorandfoundation/puya/pull/225/files#diff-0457a536f94ff2f407b843b52ba75b1e22c2482c4714d009cd2dc7a9156c45f1R585 

https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/v10/#extract3

Reply:
To replicate it in puya you will need to use int literals not UInt64,  e.g.

`
op.extract(some_bytes, 5, 0)
`

Probably will need a seperate method for this edge case

Reply:
Some tests for the edge case are added in this commit 267f745e35d81a35e41fe6ab2376ece5a4ce953a.